File: arrays_and_strings/squares_of_a_sorted_array.java
public int[] sortedSquaresDS1(int[] nums)
{
    //! @details In Java, Arrays.sort() is implemented as quicksort variant
    //!          whose space complexity is O(log N)
    //!
    //!          Time complexity O(N log N)
    //!          Space complexity O(log N) if output array is ignored

    int   N   = nums.length;
    int[] ans = new int[N];

    for (int i = 0; i < N; ++i)
        ans[i] = nums[i] * nums[i];
    
    Arrays.sort(ans);
    return ans;
}

public int[] sortedSquaresDS2(int[] nums)
{
    int   N   = nums.length;
    int[] ans = new int[N];

    int left  = 0;
    int right = N - 1;

    for (int i = N - 1; i >= 0; --i)
    {
        int square;
        if (Math.abs(nums[left]) < Math.abs(nums[right]))
        {
            square = nums[right--];
        }
        else
        {
            square = nums[left++];
        }
        ans[i] = square * square;
    }

    return ans;
}

File: arrays_and_strings/squares_of_a_sorted_array.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var sortedSquaresDS1 = function(nums) {
    const N   = nums.length;
    let   ans = [];

    for (let i = 0; i < N; i++)
    {
        ans.push(nums[i] * nums[i]);
    }

    ans.sort((a, b) => { return a - b });
    return ans;
};

let sortedSquaresDS2 = (nums) => {
    const N   = nums.length;
    const ans = [];

    let left  = 0;
    let right = N - 1;

    for (let i = N - 1; i >= 0; i--)
    {
        let square = 0;
        if (Math.abs(nums[left]) < Math.abs(nums[right]))
        {
            square = nums[right--];
        }
        else
        {
            square = nums[left++];
        }
        ans[i] = square * square;
    }

    return ans;
};